### ISSUE TITLE
Can't set an icon properly

### ISSUE THREAD

**Issue by TENX-S** [emilk/egui#920]
```rust
fn main() -> Result<()> {
 
    let mut icon_bytes = Vec::new();
    let mut img_reader = BufReader::new(File::open("asset/test.jpg")?);
    img_reader.read_to_end(&mut icon_bytes)?;

    let options = eframe::NativeOptions {
        transparent: true,
        icon_data: Some(IconData{
            rgba: icon_bytes,
            width: 32,
            height: 32,
        }),
        ..Default::default()
    };

    eframe::run_native(Box::new(HelloWorldApp::new()?), options);
}
```
The is no change after add `icon_data` to `options`

**Comment by manokara** [comment]
The `rgba` field expects raw pixel data, which is why you need to set width and height. You need to decode the JPG first using the [image crate](https://docs.rs/image/latest/image/#high-level-api).

**Comment by TENX-S** [comment]
Still not working...
```rust
fn main() -> Result<()> {
    let mut icon_bytes = Vec::new();
    let img = image::open("test.png")?;
    img.write_to(&mut icon_bytes, image::ImageFormat::Png)?;

    let options = eframe::NativeOptions {
         icon_data: Some(IconData {
            rgba: icon_bytes,
             width: 32,
             height: 32,
         }),
        ..Default::default()
    };

    eframe::run_native(Box::new(HelloWorldApp::new()), options);
}
```
test.png:
![test.png](https://user-images.githubusercontent.com/40336192/144209318-b57873bb-8086-4690-b47b-885ab3234f94.png)

**Comment by mushogenshin** [comment]
Same thing happened to me, I tried setting icon data on both Windows and MacOS but to no avail.

**Comment by ubearline** [comment]
Same issue

**Comment by overburner** [comment]
it works:
```rust
fn main() {
    let app = egui_template::WrapperApp::default();
    
    let icon = image::open("test.png").expect("Failed to open icon path").to_rgba8();
    let (icon_width, icon_height) = icon.dimensions();
    
    let options = eframe::NativeOptions {
        icon_data: Some(eframe::epi::IconData {
            rgba: icon.into_raw(),
            width: icon_width,
            height: icon_height,
        }),
        ..Default::default()
    };

    eframe::run_native(Box::new(app), options);
}
```

**Comment by MolotovCherry** [comment]
For anyone that uses Windows and wants to take the icon from the exe that they set with winres, here's a beefy function that's compatible with egui

```rust
pub fn load_app_icon() -> IconData {
    let (mut buffer, width, height) = unsafe {
        let h_instance = GetModuleHandleW(None).expect("Failed to get HINSTANCE");
        let icon = LoadImageW(
            h_instance,
            w!("corro"),
            IMAGE_ICON,
            512,
            512,
            LR_DEFAULTCOLOR,
        )
        .expect("Failed to load icon");

        let mut icon_info = ICONINFO::default();
        let res = GetIconInfo(HICON(icon.0), &mut icon_info as *mut _).as_bool();
        if !res {
            panic!("Failed to load icon info");
        }

        let mut bitmap = BITMAP::default();
        GetObjectA(
            icon_info.hbmColor,
            std::mem::size_of::<BITMAP>() as i32,
            Some(&mut bitmap as *mut _ as *mut _),
        );

        let width = bitmap.bmWidth;
        let height = bitmap.bmHeight;

        // Do not drop this Vec, we will handle dropping on our own
        let b_size = (width * height * 4) as usize;
        let mut buffer = ManuallyDrop::new(Vec::<u8>::with_capacity(b_size));
        let (b_ptr, b_cap) = (buffer.as_mut_ptr(), buffer.capacity());
        // prevent creating 2 owned buffers which would violate Rust's exclusive mutable no-aliasing rules
        std::mem::drop(buffer);

        let h_dc = CreateCompatibleDC(None);
        let h_bitmap = SelectObject(h_dc, icon_info.hbmColor);

        let mut bitmap_info = BITMAPINFO::default();
        bitmap_info.bmiHeader.biSize = std::mem::size_of::<BITMAPINFOHEADER>() as u32;
        bitmap_info.bmiHeader.biWidth = width;
        bitmap_info.bmiHeader.biHeight = height;
        bitmap_info.bmiHeader.biPlanes = 1;
        bitmap_info.bmiHeader.biBitCount = 32;
        bitmap_info.bmiHeader.biCompression = BI_RGB;
        bitmap_info.bmiHeader.biSizeImage = 0;

        let res = GetDIBits(
            h_dc,
            icon_info.hbmColor,
            0,
            height as u32,
            Some(b_ptr as *mut _),
            &mut bitmap_info as *mut _,
            DIB_RGB_COLORS,
        );
        if res == 0 {
            panic!("Failed to get RGB DI bits");
        }

        SelectObject(h_dc, h_bitmap);
        DeleteDC(h_dc);

        assert_eq!(
            bitmap_info.bmiHeader.biSizeImage as usize, b_cap,
            "returned biSizeImage must equal rgba cap size"
        );

        // retrieve the buffer
        let buffer = Vec::from_raw_parts(b_ptr, b_cap, b_cap);

        (buffer, width as u32, height as u32)
    };

    // RGBA -> ARGB
    for pixel in buffer.as_mut_slice().chunks_mut(4) {
        pixel.rotate_right(1);
    }

    // Flip the image vertically
    let row_size = width as usize * 4; // number of pixels in each row
    let row_count = buffer.len() as usize / row_size; // number of rows in the image
    for row in 0..row_count / 2 {
        // loop through half of the rows
        let start = row * row_size; // index of the start of the current row
        let end = (row_count - row - 1) * row_size; // index of the end of the current row
        for i in 0..row_size {
            buffer.swap(start + i, end + i);
        }
    }

    IconData {
        rgba: buffer,
        width: width,
        height: height,
    }
}
```

**Comment by TheMaverickProgrammer** [comment]
> 

what are your packages? I'm using winres but do not have access to the functions in your proposed routine.

**Comment by MolotovCherry** [comment]
> @MolotovCherry
> 
> what are your packages? I'm using winres but do not have access to the functions in your proposed routine.

This is windows-rs. Here's my cargo.toml, though there's a little too much in my list, but it'll work
```toml
[dependencies.windows]
version = "0.43.0"
features = [
    "Win32_UI_WindowsAndMessaging",
    "Win32_Foundation",
    "Win32_System_LibraryLoader",
    "Win32_Graphics_Gdi",
    "Win32_UI_Input_KeyboardAndMouse",
    "Win32_System_Threading",
    "Win32_UI_Controls",
    "Win32_Graphics_Dwm",
    "Win32_UI_Shell"
]
```

**Comment by TheMaverickProgrammer** [comment]
@MolotovCherry your toml file looks strange to me. Here's my attempt to fix it and cargo it's liking either:
```
[dependencies.windows]
windows-rs = {
    version = "0.43.0",
    features = [
        "Win32_UI_WindowsAndMessaging",
        "Win32_Foundation",
        "Win32_System_LibraryLoader",
        "Win32_Graphics_Gdi",
        "Win32_UI_Input_KeyboardAndMouse",
        "Win32_System_Threading",
        "Win32_UI_Controls",
        "Win32_Graphics_Dwm",
        "Win32_UI_Shell"
    ]
}
```

**Comment by MolotovCherry** [comment]
> @MolotovCherry your toml file looks strange to me. Here's my attempt to fix it and cargo it's liking either:
> 
> ```toml
> [dependencies.windows]
> windows_rs = {
>     version = "0.43.0",
>     features = [
>         "Win32_UI_WindowsAndMessaging",
>         "Win32_Foundation",
>         "Win32_System_LibraryLoader",
>         "Win32_Graphics_Gdi",
>         "Win32_UI_Input_KeyboardAndMouse",
>         "Win32_System_Threading",
>         "Win32_UI_Controls",
>         "Win32_Graphics_Dwm",
>         "Win32_UI_Shell"
>     ]
> }
> ```

This is the standard toml format recommended by windows-rs and Rust
https://github.com/microsoft/windows-rs#windows-sys
https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#choosing-features

**Comment by TheMaverickProgrammer** [comment]
@MolotovCherry thank you. I had to update rust.
After getting the dependencies to load I could not get the app to run:
```thread 'main' panicked at 'Failed to load icon: Error { code: HRESULT(0x80070715), message: "The specified resource type cannot be found in the image file." }', src\main.rs:39:10
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace```

I made sure the icon name was changed and the dimensions were correct. I'm also surprised why we have to re-specify the name and dimensions even with the `build.rs` compiling the resource file for us. Is there a way to pull that automatically from the resource file? Either way, I'm unsure why I cannot get the icon to load. BTW it is a `.ICO` file. Any ideas?

**Comment by MolotovCherry** [comment]
> @MolotovCherry thank you. I had to update rust. After getting the dependencies to load I could not get the app to run:
> 
> `thread 'main' panicked at 'Failed to load icon: Error { code: HRESULT(0x80070715), message: "The specified resource type cannot be found in the image file." }', src\main.rs:39:10` `note: run with RUST_BACKTRACE=1 environment variable to display a backtrace`
> 
> I made sure the icon name was changed and the dimensions were correct. I'm also surprised why we have to re-specify the name and dimensions even with the `build.rs` compiling the resource file for us. Is there a way to pull that automatically from the resource file? Either way, I'm unsure why I cannot get the icon to load. BTW it is a `.ICO` file. Any ideas?

Did you define the icon like so?
`res.set_icon_with_id("file.ico", "your-icon-name-here");`

**Comment by TheMaverickProgrammer** [comment]
That made the app load! but the icon is broken. smh

**Comment by MolotovCherry** [comment]
> That made the app load! but the icon is broken. smh

Need more details, e.g OS, icon, etc. But I've only tested this on Win11, so that might be a good thing to note.

**Comment by TheMaverickProgrammer** [comment]
It's a .ICO file. I think that might be why. I wanted transparency for the icon. I'm on Windows 10 as well.

**Comment by MolotovCherry** [comment]
> It's a .ICO file. I think that might be why. I wanted transparency for the icon. I'm on Windows 10 as well.

@TheMaverickProgrammer 
ICO files should have no problem. Mine is also ICO (with transparency too)

I have this to load it in (should be fairly easy)
```rust
let options = NativeOptions {
        icon_data: Some(load_app_icon()),
        ..Default::default()
    };
```
Maybe you can send your ICO file and I'll see if it works.

**Comment by TheMaverickProgrammer** [comment]
yeah I'm unsure why. my ico editor seems to like it just fine.

[favicon.zip](https://github.com/emilk/egui/files/10299272/favicon.zip)

**Comment by MolotovCherry** [comment]
> yeah I'm unsure why. my ico editor seems to like it just fine.
> 
> [favicon.zip](https://github.com/emilk/egui/files/10299272/favicon.zip)

@TheMaverickProgrammer 
Please replace `pixel.rotate_right(1);` with `pixel.swap(0, 2);` and see if that works

**Comment by TheMaverickProgrammer** [comment]
@MolotovCherry that was quick! I found that it had something to do with the color order as well but I'm still not sure what's up.

**Comment by MolotovCherry** [comment]
> @MolotovCherry that was quick! I found that it had something to do with the color order as well but I'm still not sure what's up. Thank you!!

@TheMaverickProgrammer I tested out a rainbow favicon now and the colors all display correctly. Let me know if your icon is good now

**Comment by TheMaverickProgrammer** [comment]
@MolotovCherry Didn't mean to leave you hanging. I thought it was clear that you fixed the color problem before me! Thank you!

**Comment by anilbey** [comment]
Update: eframe::epi is private. The following worked for me.

```rust
    let icon = image::open("static/icon.png")
        .expect("Failed to open icon path")
        .to_rgba8();
    let (icon_width, icon_height) = icon.dimensions();

    let options = eframe::NativeOptions {
        icon_data: Some(eframe::IconData {
            rgba: icon.into_raw(),
            width: icon_width,
            height: icon_height,
        }),
        drag_and_drop_support: true,
        ..Default::default()
    };

    eframe::run_native(
        "App Name",
        options,
        Box::new(|_cc| Box::<NWBView>::default()),
    );

```

**Comment by c-git** [comment]
> ```rust
>     let icon = image::open("static/icon.png")
> ```

For anyone coming after me, I just wanted to point out that you need to have the [image](https://crates.io/crates/image) crate as a dependency to be able to use this code. After I used `cargo add image` then the code worked for me. It might already be obvious but it wasn't to me. I mean with a bit of guessing I got there but hoping I can help the next person.

**Comment by electrorayer** [comment]
For anyone coming after me, I had to set the `viewport` field and use the `with_icon()` method to set an image.

main.rs
```rust
    let icon = image::open("icon.png").expect("Failed to open icon path").to_rgba8();
    let (icon_width, icon_height) = icon.dimensions();

    let options = eframe::NativeOptions {
        viewport: egui::ViewportBuilder::default().with_icon(egui::IconData { 
                rgba: icon.into_raw(), 
                width: icon_width, 
                height: icon_height,
        }),
        ..Default::default()
    };

    eframe::run_native(
        "MyApp", 
        options, 
        Box::new(|_| { 
            Box::<App>::default() 
        }),
    )
```

Cargo.toml

```rust
[dependencies]
eframe = "0.25.0"
egui = "0.25.0"
image = "0.24.8"

```

**Comment by c-git** [comment]
Similar example code got merged into the official template. It just reuses one of the existing image assets because it's just an example. Hopefully that will help those in the future to not need to end up googling this like how we got here

Link to the example: https://github.com/emilk/eframe_template/blob/50ce36a17201b32269bcd829bade159f923ef2aa/src/main.rs#L15

Full disclosure I created the PR but I'd have recommended it anyway as it's a full runnable example.
